File: clients/go-sdk/client/subscriptions.go
package client

import (
	"context"
	"fmt"
)

// SubscriptionListResponse — GET /api/subscriptions.
type SubscriptionListResponse struct {
	Subscriptions []SubscriptionResponse `json:"subscriptions"`
	Total         uint64                 `json:"total,omitempty"`
}

// ConfigEntry is a per-subscription custom config key/value.
type ConfigEntry struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// EventTypeBinding pairs an event type code/pattern with an optional
// filter expression. Patterns support `*` wildcards per segment.
type EventTypeBinding struct {
	EventTypeCode string `json:"eventTypeCode"`
	Filter        string `json:"filter,omitempty"`
}

// CreateSubscriptionRequest — POST /api/subscriptions.
type CreateSubscriptionRequest struct {
	Code             string             `json:"code"`
	Name             string             `json:"name"`
	Description      string             `json:"description,omitempty"`
	Endpoint         string             `json:"endpoint"`
	ConnectionID     string             `json:"connectionId,omitempty"`
	EventTypes       []EventTypeBinding `json:"eventTypes,omitempty"`
	ClientID         string             `json:"clientId,omitempty"`
	DispatchPoolID   string             `json:"dispatchPoolId,omitempty"`
	ServiceAccountID string             `json:"serviceAccountId,omitempty"`
	Mode             string             `json:"mode,omitempty"`
	TimeoutSeconds   *uint32            `json:"timeoutSeconds,omitempty"`
	MaxRetries       *uint32            `json:"maxRetries,omitempty"`
	DataOnly         bool               `json:"dataOnly,omitempty"`
}

// UpdateSubscriptionRequest — PUT /api/subscriptions/{id}.
type UpdateSubscriptionRequest struct {
	Name           *string `json:"name,omitempty"`
	Description    *string `json:"description,omitempty"`
	Endpoint       *string `json:"endpoint,omitempty"`
	ConnectionID   *string `json:"connectionId,omitempty"`
	TimeoutSeconds *uint32 `json:"timeoutSeconds,omitempty"`
	MaxRetries     *uint32 `json:"maxRetries,omitempty"`
}

// SubscriptionResponse is the platform's subscription representation.
type SubscriptionResponse struct {
	ID                string             `json:"id"`
	Code              string             `json:"code"`
	Name              string             `json:"name"`
	Description       string             `json:"description,omitempty"`
	ClientID          string             `json:"clientId,omitempty"`
	ClientIdentifier  string             `json:"clientIdentifier,omitempty"`
	EventTypes        []EventTypeBinding `json:"eventTypes,omitempty"`
	Endpoint          string             `json:"endpoint"`
	ConnectionID      string             `json:"connectionId,omitempty"`
	Queue             string             `json:"queue,omitempty"`
	Source            string             `json:"source,omitempty"`
	Status            string             `json:"status"`
	MaxAgeSeconds     uint32             `json:"maxAgeSeconds,omitempty"`
	DispatchPoolID    string             `json:"dispatchPoolId,omitempty"`
	DispatchPoolCode  string             `json:"dispatchPoolCode,omitempty"`
	DelaySeconds      uint32             `json:"delaySeconds,omitempty"`
	Sequence          int32              `json:"sequence,omitempty"`
	Mode              string             `json:"mode"`
	TimeoutSeconds    uint32             `json:"timeoutSeconds,omitempty"`
	MaxRetries        uint32             `json:"maxRetries,omitempty"`
	ServiceAccountID  string             `json:"serviceAccountId,omitempty"`
	DataOnly          bool               `json:"dataOnly,omitempty"`
	ApplicationCode   string             `json:"applicationCode,omitempty"`
	ClientScoped      bool               `json:"clientScoped,omitempty"`
	CustomConfig      []ConfigEntry      `json:"customConfig,omitempty"`
	CreatedAt         string             `json:"createdAt"`
	UpdatedAt         string             `json:"updatedAt"`
}

// SyncSubscriptionItem matches the platform's SyncSubscriptionInput.
type SyncSubscriptionItem struct {
	Code             string             `json:"code"`
	Name             string             `json:"name"`
	Description      string             `json:"description,omitempty"`
	Target           string             `json:"target"`
	ConnectionID     string             `json:"connectionId,omitempty"`
	EventTypes       []EventTypeBinding `json:"eventTypes"`
	DispatchPoolCode string             `json:"dispatchPoolCode,omitempty"`
	Mode             string             `json:"mode,omitempty"`
	MaxRetries       *uint32            `json:"maxRetries,omitempty"`
	TimeoutSeconds   *uint32            `json:"timeoutSeconds,omitempty"`
	DataOnly         bool               `json:"dataOnly,omitempty"`
}

// SyncSubscriptionsRequest — body for the per-app sync endpoint.
type SyncSubscriptionsRequest struct {
	Subscriptions []SyncSubscriptionItem `json:"subscriptions"`
}

// SubscriptionsResource — /api/subscriptions/*.
type SubscriptionsResource struct {
	c *FlowCatalystClient
}

// Create — POST /api/subscriptions.
func (r *SubscriptionsResource) Create(ctx context.Context, req *CreateSubscriptionRequest) (*SubscriptionResponse, error) {
	var out SubscriptionResponse
	if err := r.c.Post(ctx, "/api/subscriptions", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/subscriptions/{id}.
func (r *SubscriptionsResource) Get(ctx context.Context, id string) (*SubscriptionResponse, error) {
	var out SubscriptionResponse
	if err := r.c.Get(ctx, "/api/subscriptions/"+id, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List — GET /api/subscriptions?clientId=&status=.
func (r *SubscriptionsResource) List(ctx context.Context, clientID, status string) (*SubscriptionListResponse, error) {
	q := EncodeQuery("clientId", clientID, "status", status)
	var out SubscriptionListResponse
	if err := r.c.Get(ctx, "/api/subscriptions"+q, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update — PUT /api/subscriptions/{id}.
func (r *SubscriptionsResource) Update(ctx context.Context, id string, req *UpdateSubscriptionRequest) (*SubscriptionResponse, error) {
	var out SubscriptionResponse
	if err := r.c.Put(ctx, "/api/subscriptions/"+id, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Pause — POST /api/subscriptions/{id}/pause.
func (r *SubscriptionsResource) Pause(ctx context.Context, id string) error {
	return r.c.Post(ctx, "/api/subscriptions/"+id+"/pause", nil, nil)
}

// Resume — POST /api/subscriptions/{id}/resume.
func (r *SubscriptionsResource) Resume(ctx context.Context, id string) error {
	return r.c.Post(ctx, "/api/subscriptions/"+id+"/resume", nil, nil)
}

// Delete — DELETE /api/subscriptions/{id}.
func (r *SubscriptionsResource) Delete(ctx context.Context, id string) error {
	return r.c.Delete(ctx, "/api/subscriptions/"+id, nil)
}

// Sync — POST /api/applications/{appCode}/subscriptions/sync.
func (r *SubscriptionsResource) Sync(ctx context.Context, appCode string, req *SyncSubscriptionsRequest, removeUnlisted bool) (*SyncResult, error) {
	q := ""
	if removeUnlisted {
		q = "?removeUnlisted=true"
	}
	var out SyncResult
	path := fmt.Sprintf("/api/applications/%s/subscriptions/sync%s", appCode, q)
	if err := r.c.Post(ctx, path, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/subscriptions.rs
//! Subscription management operations.

use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// Paginated list of subscriptions — `GET /api/subscriptions`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionListResponse {
    pub subscriptions: Vec<SubscriptionResponse>,
    #[serde(default)]
    pub total: u64,
}

/// Custom config entry returned on subscription responses.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigEntry {
    pub key: String,
    pub value: String,
}

/// Request to create a subscription.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateSubscriptionRequest {
    /// Unique code for this subscription
    pub code: String,
    /// Human-readable name
    pub name: String,
    /// Optional description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Webhook endpoint URL
    pub endpoint: String,
    /// Connection ID (references msg_connections, optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub connection_id: Option<String>,
    /// Event type bindings (patterns with optional filters)
    #[serde(default)]
    pub event_types: Vec<EventTypeBinding>,
    /// Client ID for multi-tenant scoping
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_id: Option<String>,
    /// Dispatch pool ID for rate limiting
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dispatch_pool_id: Option<String>,
    /// Service account ID for authentication
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_account_id: Option<String>,
    /// Dispatch mode
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<String>,
    /// Webhook timeout in seconds
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<u32>,
    /// Maximum retry attempts
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_retries: Option<u32>,
    /// Send raw data only (no envelope)
    #[serde(default)]
    pub data_only: bool,
}

/// Event type binding with wildcard pattern support.
///
/// Supports patterns like `"orders:*:*:*"` to match all events from the orders app.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EventTypeBinding {
    /// Event type code or pattern (supports `*` wildcard per segment)
    pub event_type_code: String,
    /// Optional filter expression
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter: Option<String>,
}

/// Request to update a subscription.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSubscriptionRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub connection_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_retries: Option<u32>,
}

/// Subscription response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionResponse {
    pub id: String,
    pub code: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub client_id: Option<String>,
    #[serde(default)]
    pub client_identifier: Option<String>,
    #[serde(default)]
    pub event_types: Vec<EventTypeBinding>,
    pub endpoint: String,
    #[serde(default)]
    pub connection_id: Option<String>,
    #[serde(default)]
    pub queue: Option<String>,
    #[serde(default)]
    pub source: Option<String>,
    pub status: String,
    #[serde(default)]
    pub max_age_seconds: u32,
    #[serde(default)]
    pub dispatch_pool_id: Option<String>,
    #[serde(default)]
    pub dispatch_pool_code: Option<String>,
    #[serde(default)]
    pub delay_seconds: u32,
    #[serde(default)]
    pub sequence: i32,
    pub mode: String,
    #[serde(default)]
    pub timeout_seconds: u32,
    #[serde(default)]
    pub max_retries: u32,
    #[serde(default)]
    pub service_account_id: Option<String>,
    #[serde(default)]
    pub data_only: bool,
    #[serde(default)]
    pub application_code: Option<String>,
    #[serde(default)]
    pub client_scoped: bool,
    /// Per-subscription custom configuration entries (key/value pairs).
    /// Returned by the platform as part of every subscription response.
    #[serde(default)]
    pub custom_config: Vec<ConfigEntry>,
    pub created_at: String,
    pub updated_at: String,
}

/// Request body for the per-resource sync endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncSubscriptionsRequest {
    pub subscriptions: Vec<SyncSubscriptionItem>,
}

/// A subscription item for sync — matches platform's SyncSubscriptionInput.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncSubscriptionItem {
    pub code: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Webhook endpoint URL
    pub target: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub connection_id: Option<String>,
    pub event_types: Vec<SyncEventTypeBinding>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dispatch_pool_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_retries: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<u32>,
    #[serde(default)]
    pub data_only: bool,
}

/// Event type binding for subscription sync.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncEventTypeBinding {
    pub event_type_code: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter: Option<String>,
}

/// Subscriptions resource accessor — created via [`FlowCatalystClient::subscriptions`].
pub struct Subscriptions<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl Subscriptions<'_> {
    /// Create a new subscription.
    pub async fn create(
        &self,
        req: &CreateSubscriptionRequest,
    ) -> Result<SubscriptionResponse, ClientError> {
        self.client.post("/api/subscriptions", req).await
    }

    /// Get a subscription by ID.
    pub async fn get(&self, id: &str) -> Result<SubscriptionResponse, ClientError> {
        self.client.get(&format!("/api/subscriptions/{}", id)).await
    }

    /// List subscriptions with optional filters.
    pub async fn list(
        &self,
        client_id: Option<&str>,
        status: Option<&str>,
    ) -> Result<SubscriptionListResponse, ClientError> {
        let mut params = Vec::new();
        if let Some(cid) = client_id {
            params.push(format!("client_id={}", cid));
        }
        if let Some(s) = status {
            params.push(format!("status={}", s));
        }

        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };

        self.client
            .get(&format!("/api/subscriptions{}", query))
            .await
    }

    /// Update a subscription.
    pub async fn update(
        &self,
        id: &str,
        req: &UpdateSubscriptionRequest,
    ) -> Result<SubscriptionResponse, ClientError> {
        self.client
            .put(&format!("/api/subscriptions/{}", id), req)
            .await
    }

    /// Pause a subscription.
    pub async fn pause(&self, id: &str) -> Result<(), ClientError> {
        self.client
            .post_empty(&format!("/api/subscriptions/{}/pause", id))
            .await
    }

    /// Resume a subscription.
    pub async fn resume(&self, id: &str) -> Result<(), ClientError> {
        self.client
            .post_empty(&format!("/api/subscriptions/{}/resume", id))
            .await
    }

    /// Delete a subscription.
    pub async fn delete(&self, id: &str) -> Result<(), ClientError> {
        self.client
            .delete_req(&format!("/api/subscriptions/{}", id))
            .await
    }

    /// Sync subscriptions for an application — declarative reconciliation
    /// against `POST /api/applications/{appCode}/subscriptions/sync`.
    pub async fn sync(
        &self,
        app_code: &str,
        req: &SyncSubscriptionsRequest,
        remove_unlisted: bool,
    ) -> Result<crate::client::SyncResult, ClientError> {
        let query = if remove_unlisted {
            "?removeUnlisted=true"
        } else {
            ""
        };
        self.client
            .post(
                &format!(
                    "/api/applications/{}/subscriptions/sync{}",
                    app_code, query
                ),
                req,
            )
            .await
    }
}
